File: src/kitchen_system.py
#!/usr/bin/env python3
"""
Real-time Food Order Fulfillment System

This system manages food orders in a delivery-only kitchen with concurrent
order placement and pickup operations.
"""

import time
import threading
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from enum import Enum
from datetime import datetime


class Temperature(Enum):
    """Food storage temperature requirements."""
    HOT = "hot"
    COLD = "cold"
    ROOM = "room"


class ActionType(Enum):
    """Types of kitchen actions."""
    PLACE = "place"
    MOVE = "move"
    PICKUP = "pickup"
    DISCARD = "discard"


@dataclass
class FoodOrder:
    """Represents a food order with all required attributes."""
    id: str
    name: str
    temperature: Temperature
    freshness: int  # Duration in seconds
    placed_at: float  # Timestamp when placed
    stored_at: float  # Timestamp when stored
    storage_location: str  # "cooler", "heater", or "shelf"
    
    def is_fresh(self, current_time: float) -> bool:
        """Check if the order is still fresh at current time."""
        elapsed_time = current_time - self.stored_at
        ideal_freshness = self.freshness
        
        # If stored at ideal temperature, use normal freshness
        # If stored at non-ideal temperature, degrade twice as quickly
        if self.storage_location == self.temperature.value:
            return elapsed_time < ideal_freshness
        else:
            return elapsed_time < (ideal_freshness / 2)


@dataclass
class KitchenAction:
    """Represents a kitchen action for the ledger."""
    timestamp: float
    order_id: str
    action_type: ActionType
    target: str  # Storage location or action description
    details: str = ""


class KitchenSystem:
    """Main kitchen system managing order fulfillment and storage."""
    
    def __init__(self):
        # Storage containers with capacity limits
        self.cooler: List[FoodOrder] = []  # Max 6 cold orders
        self.heater: List[FoodOrder] = []  # Max 6 hot orders
        self.shelf: List[FoodOrder] = []   # Max 12 orders
        
        # Order tracking
        self.orders: Dict[str, FoodOrder] = {}
        
        # Action ledger
        self.actions: List[KitchenAction] = []
        
        # Thread safety
        self.lock = threading.RLock()
        
        # Statistics
        self.stats = {
            'orders_placed': 0,
            'orders_picked_up': 0,
            'orders_discarded': 0,
            'orders_moved': 0
        }
    
    def place_order(self, order_id: str, name: str, temperature: Temperature, 
                   freshness: int) -> bool:
        """
        Place a new order in the kitchen system.
        
        Returns True if successfully placed, False otherwise.
        """
        current_time = time.time()
        
        with self.lock:
            # Create the order
            order = FoodOrder(
                id=order_id,
                name=name,
                temperature=temperature,
                freshness=freshness,
                placed_at=current_time,
                stored_at=current_time,
                storage_location=""
            )
            
            # Try to store at ideal temperature first
            if self._try_store_at_ideal_temperature(order):
                self._log_action(current_time, order_id, ActionType.PLACE, 
                               order.storage_location, f"Stored {name} at ideal temperature")
                self.orders[order_id] = order
                self.stats['orders_placed'] += 1
                return True
            
            # If ideal storage is full, try to move existing orders and place on shelf
            if self._try_make_room_and_place(order):
                self._log_action(current_time, order_id, ActionType.PLACE, 
                               order.storage_location, f"Stored {name} after making room")
                self.orders[order_id] = order
                self.stats['orders_placed'] += 1
                return True
            
            # If still can't place, discard an order and place new one
            if self._discard_and_place(order):
                self._log_action(current_time, order_id, ActionType.PLACE, 
                               order.storage_location, f"Stored {name} after discarding old order")
                self.orders[order_id] = order
                self.stats['orders_placed'] += 1
                return True
            
            return False
    
    def _try_store_at_ideal_temperature(self, order: FoodOrder) -> bool:
        """Try to store order at its ideal temperature."""
        if order.temperature == Temperature.COLD and len(self.cooler) < 6:
            self.cooler.append(order)
            order.storage_location = "cooler"
            return True
        elif order.temperature == Temperature.HOT and len(self.heater) < 6:
            self.heater.append(order)
            order.storage_location = "heater"
            return True
        elif order.temperature == Temperature.ROOM and len(self.shelf) < 12:
            self.shelf.append(order)
            order.storage_location = "shelf"
            return True
        return False
    
    def _try_make_room_and_place(self, order: FoodOrder) -> bool:
        """Try to move existing orders to make room for new order."""
        # Try to move cold orders from shelf to cooler
        if order.temperature == Temperature.COLD and len(self.cooler) < 6:
            moved = self._move_from_shelf_to_cooler()
            if moved and len(self.shelf) < 12:
                self.shelf.append(order)
                order.storage_location = "shelf"
                return True
        
        # Try to move hot orders from shelf to heater
        elif order.temperature == Temperature.HOT and len(self.heater) < 6:
            moved = self._move_from_shelf_to_heater()
            if moved and len(self.shelf) < 12:
                self.shelf.append(order)
                order.storage_location = "shelf"
                return True
        
        # Try to place directly on shelf if there's room
        elif len(self.shelf) < 12:
            self.shelf.append(order)
            order.storage_location = "shelf"
            return True
        
        return False
    
    def _move_from_shelf_to_cooler(self) -> bool:
        """Move a cold order from shelf to cooler if possible."""
        for i, order in enumerate(self.shelf):
            if order.temperature == Temperature.COLD and len(self.cooler) < 6:
                moved_order = self.shelf.pop(i)
                self.cooler.append(moved_order)
                moved_order.storage_location = "cooler"
                
                current_time = time.time()
                self._log_action(current_time, moved_order.id, ActionType.MOVE, 
                               "cooler", f"Moved {moved_order.name} from shelf to cooler")
                self.stats['orders_moved'] += 1
                return True
        return False
    
    def _move_from_shelf_to_heater(self) -> bool:
        """Move a hot order from shelf to heater if possible."""
        for i, order in enumerate(self.shelf):
            if order.temperature == Temperature.HOT and len(self.heater) < 6:
                moved_order = self.shelf.pop(i)
                self.heater.append(moved_order)
                moved_order.storage_location = "heater"
                
                current_time = time.time()
                self._log_action(current_time, moved_order.id, ActionType.MOVE, 
                               "heater", f"Moved {moved_order.name} from shelf to heater")
                self.stats['orders_moved'] += 1
                return True
        return False
    
    def _discard_and_place(self, order: FoodOrder) -> bool:
        """Discard an order from shelf to make room for new order."""
        if len(self.shelf) >= 12:
            # Choose order to discard based on freshness and temperature mismatch
            discard_index = self._choose_order_to_discard()
            if discard_index is not None:
                discarded_order = self.shelf.pop(discard_index)
                current_time = time.time()
                
                self._log_action(current_time, discarded_order.id, ActionType.DISCARD, 
                               "shelf", f"Discarded {discarded_order.name} to make room")
                self.stats['orders_discarded'] += 1
                
                # Now place the new order
                self.shelf.append(order)
                order.storage_location = "shelf"
                return True
        
        return False
    
    def _choose_order_to_discard(self) -> Optional[int]:
        """
        Choose which order to discard when shelf is full.
        
        Strategy: Prioritize orders that are:
        1. Already expired/freshness exceeded
        2. Stored at non-ideal temperature (degrading faster)
        3. Closest to expiration
        """
        current_time = time.time()
        best_discard_index = None
        best_discard_score = float('-inf')
        
        for i, order in enumerate(self.shelf):
            # Calculate discard score (higher = better to discard)
            score = 0
            
            # Check if expired
            if not order.is_fresh(current_time):
                score += 1000  # High priority to discard expired orders
            
            # Check temperature mismatch penalty
            if order.storage_location != order.temperature.value:
                score += 500  # Medium priority for temperature mismatches
            
            # Add time-based score (closer to expiration = higher score)
            elapsed_time = current_time - order.stored_at
            ideal_freshness = order.freshness
            if order.storage_location == order.temperature.value:
                time_ratio = elapsed_time / ideal_freshness
            else:
                time_ratio = (elapsed_time * 2) / ideal_freshness
            
            score += int(time_ratio * 100)
            
            if score > best_discard_score:
                best_discard_score = score
                best_discard_index = i
        
        return best_discard_index
    
    def pickup_order(self, order_id: str) -> Optional[FoodOrder]:
        """
        Pick up an order for delivery.
        
        Returns the order if successful, None if not found or expired.
        """
        current_time = time.time()
        
        with self.lock:
            if order_id not in self.orders:
                return None
            
            order = self.orders[order_id]
            
            # Check if order is still fresh
            if not order.is_fresh(current_time):
                # Order expired, discard it
                self._remove_order_from_storage(order)
                self._log_action(current_time, order_id, ActionType.DISCARD, 
                               order.storage_location, f"Discarded expired {order.name}")
                self.stats['orders_discarded'] += 1
                del self.orders[order_id]
                return None
            
            # Remove order from storage
            self._remove_order_from_storage(order)
            self._log_action(current_time, order_id, ActionType.PICKUP, 
                           order.storage_location, f"Picked up {order.name}")
            self.stats['orders_picked_up'] += 1
            
            # Remove from tracking
            del self.orders[order_id]
            
            return order
    
    def _remove_order_from_storage(self, order: FoodOrder):
        """Remove order from its current storage location."""
        if order.storage_location == "cooler":
            self.cooler = [o for o in self.cooler if o.id != order.id]
        elif order.storage_location == "heater":
            self.heater = [o for o in self.heater if o.id != order.id]
        elif order.storage_location == "shelf":
            self.shelf = [o for o in self.shelf if o.id != order.id]
    
    def _log_action(self, timestamp: float, order_id: str, action_type: ActionType, 
                   target: str, details: str = ""):
        """Log a kitchen action to the ledger."""
        action = KitchenAction(
            timestamp=timestamp,
            order_id=order_id,
            action_type=action_type,
            target=target,
            details=details
        )
        self.actions.append(action)
        
        # Output to console in human-readable format
        dt = datetime.fromtimestamp(timestamp)
        time_str = dt.strftime("%H:%M:%S.%f")[:-3]
        print(f"[{time_str}] {action_type.value.upper()}: Order {order_id} -> {target} {details}")
    
    def get_storage_status(self) -> Dict[str, Dict]:
        """Get current storage status."""
        with self.lock:
            return {
                'cooler': {
                    'count': len(self.cooler),
                    'capacity': 6,
                    'orders': [{'id': o.id, 'name': o.name, 'temperature': o.temperature.value} 
                              for o in self.cooler]
                },
                'heater': {
                    'count': len(self.heater),
                    'capacity': 6,
                    'orders': [{'id': o.id, 'name': o.name, 'temperature': o.temperature.value} 
                              for o in self.heater]
                },
                'shelf': {
                    'count': len(self.shelf),
                    'capacity': 12,
                    'orders': [{'id': o.id, 'name': o.name, 'temperature': o.temperature.value} 
                              for o in self.shelf]
                }
            }
    
    def get_actions_ledger(self) -> List[KitchenAction]:
        """Get the complete actions ledger."""
        with self.lock:
            return self.actions.copy()
    
    def get_statistics(self) -> Dict[str, int]:
        """Get system statistics."""
        with self.lock:
            return self.stats.copy()
    
    def cleanup_expired_orders(self):
        """Remove expired orders from storage."""
        current_time = time.time()
        
        with self.lock:
            # Check cooler
            expired_cooler = [o for o in self.cooler if not o.is_fresh(current_time)]
            for order in expired_cooler:
                self.cooler.remove(order)
                self._log_action(current_time, order.id, ActionType.DISCARD, 
                               "cooler", f"Auto-discarded expired {order.name}")
                self.stats['orders_discarded'] += 1
                if order.id in self.orders:
                    del self.orders[order.id]
            
            # Check heater
            expired_heater = [o for o in self.heater if not o.is_fresh(current_time)]
            for order in expired_heater:
                self.heater.remove(order)
                self._log_action(current_time, order.id, ActionType.DISCARD, 
                               "heater", f"Auto-discarded expired {order.name}")
                self.stats['orders_discarded'] += 1
                if order.id in self.orders:
                    del self.orders[order.id]
            
            # Check shelf
            expired_shelf = [o for o in self.shelf if not o.is_fresh(current_time)]
            for order in expired_shelf:
                self.shelf.remove(order)
                self._log_action(current_time, order.id, ActionType.DISCARD, 
                               "shelf", f"Auto-discarded expired {order.name}")
                self.stats['orders_discarded'] += 1
                if order.id in self.orders:
                    del self.orders[order.id]

File: java-food-system/src/main/java/com/kitchen/service/KitchenSystem.java
package com.kitchen.service;

import com.kitchen.model.*;
import java.util.*;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.concurrent.locks.ReadWriteLock;

/**
 * Main kitchen system managing order fulfillment and storage.
 */
public class KitchenSystem {
    // Storage containers with capacity limits
    private final List<FoodOrder> cooler = new ArrayList<>();  // Max 6 cold orders
    private final List<FoodOrder> heater = new ArrayList<>();  // Max 6 hot orders
    private final List<FoodOrder> shelf = new ArrayList<>();   // Max 12 orders
    
    // Order tracking
    private final Map<String, FoodOrder> orders = new HashMap<>();
    
    // Action ledger
    private final List<KitchenAction> actions = new ArrayList<>();
    
    // Thread safety
    private final ReadWriteLock lock = new ReentrantReadWriteLock();
    
    // Statistics
    private final Map<String, Integer> stats = new HashMap<>();
    
    public KitchenSystem() {
        stats.put("orders_placed", 0);
        stats.put("orders_picked_up", 0);
        stats.put("orders_discarded", 0);
        stats.put("orders_moved", 0);
    }
    
    /**
     * Place a new order in the kitchen system.
     * 
     * @return true if successfully placed, false otherwise
     */
    public boolean placeOrder(String orderId, String name, Temperature temperature, int freshness) {
        long currentTime = System.currentTimeMillis();
        
        lock.writeLock().lock();
        try {
            // Create the order
            FoodOrder order = new FoodOrder(orderId, name, temperature, freshness, 
                                         currentTime, currentTime, "");
            
            // Try to store at ideal temperature first
            if (tryStoreAtIdealTemperature(order)) {
                logAction(currentTime, orderId, ActionType.PLACE, 
                         order.getStorageLocation(), "Stored " + name + " at ideal temperature");
                orders.put(orderId, order);
                stats.put("orders_placed", stats.get("orders_placed") + 1);
                return true;
            }
            
            // If ideal storage is full, try to move existing orders and place on shelf
            if (tryMakeRoomAndPlace(order)) {
                logAction(currentTime, orderId, ActionType.PLACE, 
                         order.getStorageLocation(), "Stored " + name + " after making room");
                orders.put(orderId, order);
                stats.put("orders_placed", stats.get("orders_placed") + 1);
                return true;
            }
            
            // If still can't place, discard an order and place new one
            if (discardAndPlace(order)) {
                logAction(currentTime, orderId, ActionType.PLACE, 
                         order.getStorageLocation(), "Stored " + name + " after discarding old order");
                orders.put(orderId, order);
                stats.put("orders_placed", stats.get("orders_placed") + 1);
                return true;
            }
            
            return false;
        } finally {
            lock.writeLock().unlock();
        }
    }
    
    private boolean tryStoreAtIdealTemperature(FoodOrder order) {
        if (order.getTemperature() == Temperature.COLD && cooler.size() < 6) {
            cooler.add(order);
            order.setStorageLocation("cooler");
            return true;
        } else if (order.getTemperature() == Temperature.HOT && heater.size() < 6) {
            heater.add(order);
            order.setStorageLocation("heater");
            return true;
        } else if (order.getTemperature() == Temperature.ROOM && shelf.size() < 12) {
            shelf.add(order);
            order.setStorageLocation("shelf");
            return true;
        }
        return false;
    }
    
    private boolean tryMakeRoomAndPlace(FoodOrder order) {
        // Try to move cold orders from shelf to cooler
        if (order.getTemperature() == Temperature.COLD && cooler.size() < 6) {
            boolean moved = moveFromShelfToCooler();
            if (moved && shelf.size() < 12) {
                shelf.add(order);
                order.setStorageLocation("shelf");
                return true;
            }
        }
        
        // Try to move hot orders from shelf to heater
        else if (order.getTemperature() == Temperature.HOT && heater.size() < 6) {
            boolean moved = moveFromShelfToHeater();
            if (moved && shelf.size() < 12) {
                shelf.add(order);
                order.setStorageLocation("shelf");
                return true;
            }
        }
        
        // Try to place directly on shelf if there's room
        else if (shelf.size() < 12) {
            shelf.add(order);
            order.setStorageLocation("shelf");
            return true;
        }
        
        return false;
    }
    
    private boolean moveFromShelfToCooler() {
        for (int i = 0; i < shelf.size(); i++) {
            FoodOrder order = shelf.get(i);
            if (order.getTemperature() == Temperature.COLD && cooler.size() < 6) {
                FoodOrder movedOrder = shelf.remove(i);
                cooler.add(movedOrder);
                movedOrder.setStorageLocation("cooler");
                
                long currentTime = System.currentTimeMillis();
                logAction(currentTime, movedOrder.getId(), ActionType.MOVE, 
                         "cooler", "Moved " + movedOrder.getName() + " from shelf to cooler");
                stats.put("orders_moved", stats.get("orders_moved") + 1);
                return true;
            }
        }
        return false;
    }
    
    private boolean moveFromShelfToHeater() {
        for (int i = 0; i < shelf.size(); i++) {
            FoodOrder order = shelf.get(i);
            if (order.getTemperature() == Temperature.HOT && heater.size() < 6) {
                FoodOrder movedOrder = shelf.remove(i);
                heater.add(movedOrder);
                movedOrder.setStorageLocation("heater");
                
                long currentTime = System.currentTimeMillis();
                logAction(currentTime, movedOrder.getId(), ActionType.MOVE, 
                         "heater", "Moved " + movedOrder.getName() + " from shelf to heater");
                stats.put("orders_moved", stats.get("orders_moved") + 1);
                return true;
            }
        }
        return false;
    }
    
    private boolean discardAndPlace(FoodOrder order) {
        if (shelf.size() >= 12) {
            // Choose order to discard based on freshness and temperature mismatch
            Integer discardIndex = chooseOrderToDiscard();
            if (discardIndex != null) {
                FoodOrder discardedOrder = shelf.remove((int) discardIndex);
                long currentTime = System.currentTimeMillis();
                
                logAction(currentTime, discardedOrder.getId(), ActionType.DISCARD, 
                         "shelf", "Discarded " + discardedOrder.getName() + " to make room");
                stats.put("orders_discarded", stats.get("orders_discarded") + 1);
                
                // Now place the new order
                shelf.add(order);
                order.setStorageLocation("shelf");
                return true;
            }
        }
        
        return false;
    }
    
    private Integer chooseOrderToDiscard() {
        /**
         * Choose which order to discard when shelf is full.
         * 
         * Strategy: Prioritize orders that are:
         * 1. Already expired/freshness exceeded
         * 2. Stored at non-ideal temperature (degrading faster)
         * 3. Closest to expiration
         */
        long currentTime = System.currentTimeMillis();
        Integer bestDiscardIndex = null;
        double bestDiscardScore = Double.NEGATIVE_INFINITY;
        
        for (int i = 0; i < shelf.size(); i++) {
            FoodOrder order = shelf.get(i);
            // Calculate discard score (higher = better to discard)
            double score = 0;
            
            // Check if expired
            if (!order.isFresh(currentTime)) {
                score += 1000; // High priority to discard expired orders
            }
            
            // Check temperature mismatch penalty
            if (!order.getStorageLocation().equals(order.getTemperature().getValue())) {
                score += 500; // Medium priority for temperature mismatches
            }
            
            // Add time-based score (closer to expiration = higher score)
            long elapsedTime = currentTime - order.getStoredAt();
            long idealFreshness = order.getFreshness();
            double timeRatio;
            if (order.getStorageLocation().equals(order.getTemperature().getValue())) {
                timeRatio = (double) elapsedTime / idealFreshness;
            } else {
                timeRatio = (double) (elapsedTime * 2) / idealFreshness;
            }
            
            score += (int) (timeRatio * 100);
            
            if (score > bestDiscardScore) {
                bestDiscardScore = score;
                bestDiscardIndex = i;
            }
        }
        
        return bestDiscardIndex;
    }
    
    /**
     * Pick up an order for delivery.
     * 
     * @return the order if successful, null if not found or expired
     */
    public FoodOrder pickupOrder(String orderId) {
        long currentTime = System.currentTimeMillis();
        
        lock.writeLock().lock();
        try {
            if (!orders.containsKey(orderId)) {
                return null;
            }
            
            FoodOrder order = orders.get(orderId);
            
            // Check if order is still fresh
            if (!order.isFresh(currentTime)) {
                // Order expired, discard it
                removeOrderFromStorage(order);
                logAction(currentTime, orderId, ActionType.DISCARD, 
                         order.getStorageLocation(), "Discarded expired " + order.getName());
                stats.put("orders_discarded", stats.get("orders_discarded") + 1);
                orders.remove(orderId);
                return null;
            }
            
            // Remove order from storage
            removeOrderFromStorage(order);
            logAction(currentTime, orderId, ActionType.PICKUP, 
                     order.getStorageLocation(), "Picked up " + order.getName());
            stats.put("orders_picked_up", stats.get("orders_picked_up") + 1);
            
            // Remove from tracking
            orders.remove(orderId);
            
            return order;
        } finally {
            lock.writeLock().unlock();
        }
    }
    
    private void removeOrderFromStorage(FoodOrder order) {
        if ("cooler".equals(order.getStorageLocation())) {
            cooler.removeIf(o -> o.getId().equals(order.getId()));
        } else if ("heater".equals(order.getStorageLocation())) {
            heater.removeIf(o -> o.getId().equals(order.getId()));
        } else if ("shelf".equals(order.getStorageLocation())) {
            shelf.removeIf(o -> o.getId().equals(order.getId()));
        }
    }
    
    private void logAction(long timestamp, String orderId, ActionType actionType, 
                          String target, String details) {
        // Log action to the ledger
        KitchenAction action = new KitchenAction(timestamp, orderId, actionType, target, details);
        actions.add(action);
        
        // Output to console in human-readable format
        System.out.printf("[%s] %s: Order %s -> %s %s%n", 
                         action.getFormattedTimestamp(), 
                         actionType.getValue().toUpperCase(), 
                         orderId, target, details);
    }
    
    /**
     * Get current storage status.
     */
    public Map<String, Map<String, Object>> getStorageStatus() {
        lock.readLock().lock();
        try {
            Map<String, Map<String, Object>> status = new HashMap<>();
            
            // Cooler status
            Map<String, Object> coolerStatus = new HashMap<>();
            coolerStatus.put("count", cooler.size());
            coolerStatus.put("capacity", 6);
            List<Map<String, String>> coolerOrders = new ArrayList<>();
            for (FoodOrder order : cooler) {
                Map<String, String> orderInfo = new HashMap<>();
                orderInfo.put("id", order.getId());
                orderInfo.put("name", order.getName());
                orderInfo.put("temperature", order.getTemperature().getValue());
                coolerOrders.add(orderInfo);
            }
            coolerStatus.put("orders", coolerOrders);
            status.put("cooler", coolerStatus);
            
            // Heater status
            Map<String, Object> heaterStatus = new HashMap<>();
            heaterStatus.put("count", heater.size());
            heaterStatus.put("capacity", 6);
            List<Map<String, String>> heaterOrders = new ArrayList<>();
            for (FoodOrder order : heater) {
                Map<String, String> orderInfo = new HashMap<>();
                orderInfo.put("id", order.getId());
                orderInfo.put("name", order.getName());
                orderInfo.put("temperature", order.getTemperature().getValue());
                heaterOrders.add(orderInfo);
            }
            heaterStatus.put("orders", heaterOrders);
            status.put("heater", heaterStatus);
            
            // Shelf status
            Map<String, Object> shelfStatus = new HashMap<>();
            shelfStatus.put("count", shelf.size());
            shelfStatus.put("capacity", 12);
            List<Map<String, String>> shelfOrders = new ArrayList<>();
            for (FoodOrder order : shelf) {
                Map<String, String> orderInfo = new HashMap<>();
                orderInfo.put("id", order.getId());
                orderInfo.put("name", order.getName());
                orderInfo.put("temperature", order.getTemperature().getValue());
                shelfOrders.add(orderInfo);
            }
            shelfStatus.put("orders", shelfOrders);
            status.put("shelf", shelfStatus);
            
            return status;
        } finally {
            lock.readLock().unlock();
        }
    }
    
    /**
     * Get the complete actions ledger.
     */
    public List<KitchenAction> getActionsLedger() {
        lock.readLock().lock();
        try {
            return new ArrayList<>(actions);
        } finally {
            lock.readLock().unlock();
        }
    }
    
    /**
     * Get system statistics.
     */
    public Map<String, Integer> getStatistics() {
        lock.readLock().lock();
        try {
            return new HashMap<>(stats);
        } finally {
            lock.readLock().unlock();
        }
    }
    
    /**
     * Remove expired orders from storage.
     */
    public void cleanupExpiredOrders() {
        long currentTime = System.currentTimeMillis();
        
        lock.writeLock().lock();
        try {
            // Check cooler
            List<FoodOrder> expiredCooler = new ArrayList<>();
            for (FoodOrder order : cooler) {
                if (!order.isFresh(currentTime)) {
                    expiredCooler.add(order);
                }
            }
            for (FoodOrder order : expiredCooler) {
                cooler.remove(order);
                logAction(currentTime, order.getId(), ActionType.DISCARD, 
                         "cooler", "Auto-discarded expired " + order.getName());
                stats.put("orders_discarded", stats.get("orders_discarded") + 1);
                if (orders.containsKey(order.getId())) {
                    orders.remove(order.getId());
                }
            }
            
            // Check heater
            List<FoodOrder> expiredHeater = new ArrayList<>();
            for (FoodOrder order : heater) {
                if (!order.isFresh(currentTime)) {
                    expiredHeater.add(order);
                }
            }
            for (FoodOrder order : expiredHeater) {
                heater.remove(order);
                logAction(currentTime, order.getId(), ActionType.DISCARD, 
                         "heater", "Auto-discarded expired " + order.getName());
                stats.put("orders_discarded", stats.get("orders_discarded") + 1);
                if (orders.containsKey(order.getId())) {
                    orders.remove(order.getId());
                }
            }
            
            // Check shelf
            List<FoodOrder> expiredShelf = new ArrayList<>();
            for (FoodOrder order : shelf) {
                if (!order.isFresh(currentTime)) {
                    expiredShelf.add(order);
                }
            }
            for (FoodOrder order : expiredShelf) {
                shelf.remove(order);
                logAction(currentTime, order.getId(), ActionType.DISCARD, 
                         "shelf", "Auto-discarded expired " + order.getName());
                stats.put("orders_discarded", stats.get("orders_discarded") + 1);
                if (orders.containsKey(order.getId())) {
                    orders.remove(order.getId());
                }
            }
        } finally {
            lock.writeLock().unlock();
        }
    }
}